src/components/tutor-dialog.cjsx

```diff
@@ -0,0 +1,97 @@
+BS = require 'react-bootstrap'
+React = require 'react'
+_ = require 'underscore'
+{Promise} = require 'es6-promise'
+
+
+DialogProperties =
+  title:     React.PropTypes.string.isRequired
+  onOk:      React.PropTypes.func.isRequired
+  onCancel:  React.PropTypes.func.isRequired
+  body:      React.PropTypes.element.isRequired
+  show:      React.PropTypes.bool
+  buttons:   React.PropTypes.arrayOf( React.PropTypes.element )
+  className: React.PropTypes.string
+
+# This is the "real" dialog component. It's rendered to a div under document.body
+DetachedTutorDialog = React.createClass
+
+  displayName: 'DetachedTutorDialog'
+
+  propTypes: DialogProperties
+
+  getInitialState: ->
+    show: true
+
+  componentWillReceiveProps: (nextProps) ->
+    @setState(show: nextProps.show) if nextProps.show?
+
+  _hide: ->
+    @setState(show: false)
+
+  hide: ->
+    @_hide()
+    @props.onCancel()
+
+  render: ->
+    return null unless @state.show
+    buttons = @props.buttons or [
+        <BS.Button key='cancel' className='cancel'
+          onClick={_.compose(@props.onCancel,  @_hide)}>Cancel</BS.Button>
+        <BS.Button key='ok'     className='ok'
+          onClick={_.compose(@props.onOk, @_hide)} bsStyle='primary'>OK</BS.Button>
+    ]
+    classes = ['tutor-dialog']
+    classes.push @props.className if @props.className
+
+    <BS.Modal className={classes.join(' ')} onRequestHide={@hide} title={@props.title}>
+      <div className='modal-body'>
+        {@props.body}
+      </div>
+      <div className='modal-footer'>
+        {buttons}
+      </div>
+    </BS.Modal>
+
+module.exports = TutorDialog = React.createClass
+  displayName: 'TutorDialog'
+
+  propTypes: _.omit(DialogProperties, 'body')
+
+  componentDidMount: ->
+    TutorDialog.show(_.extend({}, @props, body: @props.children)).then(
+      => @props.onOk?()
```

Comment:
Because (while unlikely) the props could be updated. If they were then the dialog would call the old function.  We still should pass on the result through though.  fixing that now